crates/codegen/runtime/generator/src/parser/mod.rs

```diff
@@ -56,57 +58,73 @@ struct ScannerContextModel {
+impl ParserModel {
+    pub fn from_language(language: &Rc<Language>) -> Self {
+        // First, resolve the grammar structure from the flat list of items
+        let resolved = ResolveCtx::resolve(language);
 
-    /// Defines the top-level parser functions in `Language`.
-    parser_functions: BTreeMap<Identifier, RustCode>, // (name of parser, code)
-    /// Defines the top-level trivia parser functions in `Language`.
-    trivia_parser_functions: BTreeMap<Identifier, RustCode>, // (name of parser, code)
+        // Collect all parser functions
+        let parser_fns = ParserFunctions::collect(&resolved);
+        // Collect all scanner contexts and their scanners
+        let mut acc = ScannerContextCollector::default();
+        resolved.to_grammar().accept_visitor(&mut acc);
+
+        // Combine everything into the final model
+        acc.into_model(&resolved, parser_fns)
+    }
+}
+
+#[derive(Default)]
+struct ScannerContextCollector {
+    scanner_contexts: BTreeMap<Identifier, ScannerContextCollectorState>,
 
     /// Makes sure to codegen the scanner functions that are referenced by other scanners.
     top_level_scanner_names: BTreeSet<Identifier>,
-    /// Lookup table for all scanners; used to generate trie scanners.
-    all_scanners: BTreeMap<Identifier, ScannerDefinitionRef>,
     /// The current context of a parent scanner/parser being processed.
     current_context_name: Option<Identifier>,
 }
 
 #[derive(Default)]
-struct ScannerContextAccumulatorState {
-    /// Set of delimiter pairs for this context that are used in delimited error recovery.
+struct ScannerContextCollectorState {
     delimiters: BTreeMap<Identifier, Identifier>,
     scanner_definitions: BTreeSet<Identifier>,
     keyword_scanner_defs: BTreeMap<Identifier, Rc<model::KeywordItem>>,
 }
 
-impl ParserModel {
-    pub fn from_language(language: &Rc<Language>) -> Self {
-        // First, we construct the DSLv1 model from the DSLv2 definition...
-        let grammar = Grammar::from_dsl_v2(language);
-        // ...which we then transform into the parser model
-        let mut acc = ParserAccumulatorState::default();
-        grammar.accept_visitor(&mut acc);
-
-        acc.into_model()
-    }
-}
-
-impl ParserAccumulatorState {
+impl ScannerContextCollector {
     fn set_current_context(&mut self, name: Identifier) {
         self.current_context_name = Some(name.clone());
         self.scanner_contexts.entry(name).or_default();
     }
 
-    fn current_context(&mut self) -> &mut ScannerContextAccumulatorState {
+    fn current_context(&mut self) -> &mut ScannerContextCollectorState {
         self.scanner_contexts
             .get_mut(self.current_context_name.as_ref().unwrap())
             .expect("context must be set with `set_current_context`")
     }
 
-    fn into_model(self) -> ParserModel {
+    // TODO: Separate it to a function that combines the accumulated state with the parser_fns
```

Comment:
nit: I suggest attaching an issue number to these `TODO` comments, for future upkeeping. If it is a new issue, and it is not a high priority, we can put it in the `Backlog` project for now.

Reply:
I figured that the overhead of the issue is bigger than the fix - now, we combine it in the `ParserModel::from_language` and `ScannerContextCollector::into_model` returns only the relevant fields rather than immediately combining them with the ones in `ParserFunctions`.